# Introduction to Python

Welcome to RAL's introductory course on Python, a coding language used by amateurs and experts alike across all fields of physics.  We'll guide you through the fundamentals of Python and some of its basic functions with some simple exercises.  If you're feeling confident, several challenges can be found throughout this notebook.

## 1. Getting started

### 1.1 Code cells

We can't do much coding if we can't type anywhere!  Below this paragraph is a *code cell*, where Python code can be entered.  It's currently occupied by a small line of code, but you can see the result of running this code by holding the *shift* key on your keyboard and pressing *enter*.

In [42]:
print("Welcome to Python!")

Welcome to Python!


Congratulations - you've just executed your first Python instruction!  Now run the below code cells.

In [43]:
a = 3

In [ ]:
b = 2*a

In [44]:
print(b)

2.1


In [45]:
type(b)

float

In [ ]:
a*b

In [ ]:
b = 'beep'

In [46]:
type(b)

float

What we have done is define two variables *a* and *b* and determined what data type they both are.  Let's discuss data types further.

### 1.2 Data types & operators

### 1.2.1 Numerical data types

There are several different data types used in Python, each with their own properties and affiliations.  Let's start with *integer variables*, otherwise known as *integers* or *ints*.  Like the integers in mathematics, they are positive or negative whole numbers. Below are some integer variables.

In [ ]:
1 + 1
a = 4

*Floating-point numbers* or *floats* are real numbers with floating point representation; that is, decimal points.  You may see the character *e* or *E* appear within a float to designate orders of magnitude.  Below are some floats.

In [ ]:
b = 2.1
c = 4E-7

*Complex numbers* are numbers that contain a real component and an imaginary component - in mathematics, an imaginary number is a multiple of the number *i*, which is the square root of -1 and is not seen in the macroscopic world you're used to.  Below are some complex numbers, with their real and imaginary values obtained separately.  Note that *j* rather than *i* is used to represent the imaginary number *i*.

In [47]:
d = 1.5 + 0.5j
print(d.real)
print(d.imag)

1.5
0.5


*Booleans* are not really numbers or groups of numbers at all.  They can either be *True* or *False* and are denoted by the class *bool* when we determine their type.  An example of a boolean is below.

In [48]:
test = (4 < 7)
print(test)
type(test)

True


bool

### 1.2.2 Sequenced data types

A *string* is a collection of one or more unicode characters between two ' or " marks, or even '''.  In fact, there is no character data type in Python; a character would simply be a string of length 1.  Examples of strings are below.

In [ ]:
'Beep'
"Welcome to Python!"

*Lists* are similar to strings, except the items in a list do not need to be of the same type. They are created by placing the sequence of characters between two square brackets.  Examples of lists are below.

In [ ]:
[] # This is an empty list
L = [1,2,3,4,5]

You may have noticed that the words after the dash didn't interfere with our code.  This is because writing anything after a dash means it is essentially ignored by Python.  This is very useful for annotating your code.

A useful skill is to be able to access elements of a list, as well as 'slicing' the list between two elements.  See how this is done below.

In [49]:
print(L[0])
print(L[1])
print(L[2:4])
print(L[3:])

1
2
[3, 4]
[4, 5]


It's important to remember that the *0th* element of our list is the first item rather than the *1st*.  It may seem a little confusing, but you'll get used to it.

*Tuples* are similar to lists, though the only difference is that they are immutable, meaning they cannot be changed after they are created.  An example of a tuple is below.

In [50]:
t = 1, 2, 3, 'Surprise!'
print(t[0])
print(t[3])

1
Surprise!


### 1.2.3 Arithmetic operators

In the expression *a + b*, *a* and *b* are called *operands* and *+* is called an operator.  In short, it changes the operands in some way.  There are many types of operators in Python but some of them can be a little confusing.  Because of this, we'll just cover the arithmetic and comparison operators as you'll be using them all the time.  See the table below.

|Operator|Description                                                                   |Example       |
|-       |-                                                                             |-             |
|$+$     |Addition - adds operands on either side of the operator                       |$2 + 2 = 4$   |
|$-$     |Subtraction - subtracts right-hand operand from-left hand operand             |$5 - 2 = 3$   |
|$*$     |Multiplication - multiplies values on either side of operator                 |$3*3 = 9$     |
|$/$     |Division - divides left-hand operand by right-hand operand                    |$22/8 = 22.75$|
|$**$    |Exponent - raises left-hand operand to the power of right-hand operand        |$2**3 = 8$    |
|%       |Modulo - divides left-hand operand by right-hand operand and returns remainder|13%3 = 1      |
|$//$    |Floor division - same as with the division operator, but with decimals removed|$22/8 = 2$    |

<table>
<thead>
<tr><th>Operator</th><th>Description</th><th>Example</th></tr>
</thead>
<tbody>
<tr><td>$+$</td><td>B</td><td>$2 + 2 = 4$</td></tr>
<tr><td>$-$</td><td>B</td><td>$5 - 2 = 3$</td></tr>
<tr><td>$*$</td><td>B</td><td>$3*3 = 9$</td></tr>
<tr><td>$/$</td><td>B</td><td>$22/8 = 22.75$</td></tr>
<tr><td>$**$</td><td>B</td><td>$2**3 = 8$</td></tr>
<tr><td>%</td><td>B</td><td>$13%3 = 1$</td></tr>
<tr><td>$//$</td><td>B</td><td>$22/8 = 2$</td></tr>
</tbody>
</table>

### 1.2.4 Comparison operators

Comparison operators are more case-specific than arithmetic operators, but they are still handy to know.  We'll likely encounter them later in this course. See the table below.

|Operator|Description                                                                          |Example       |
|-       |-                                                                                    |-             |
|$==$    |Equal to - compares both operands, returns *True* if they are equal                  |$2 + 2 = 4$   |
|$!=$    |Not equal to - compares both operands, returns *True* if they are not equal          |$5 - 2 = 3$   |
|$<>$    |Alternative form of the *not equal to* operator                                      |$3*3 = 9$     |
|$>$     |Greater than - returns *True* if left-hand operand is greater than right-hand operand|$22/8 = 22.75$|
|$<$     |Less than - returns *True* if left-hand operand is less than right-hand operand      |$2**3 = 8$    |
|$>=$    |Greater than or equal to - returns *True* if left-hand operand is greater than or equal to right-hand operand       |13%3 = 1      |
|$<=$    |Less than or equal to - returns *True* if left-hand operand is less than or equal to right-hand operand       |$22/8 = 2$    |

There is one *assignment* operator that we really ought to cover - the equality operator, *=*.  In short, it assigns a value from a right-hand operad to a left-hand operand.  For instance, in the expression $a = 4$, we have assigned a value of 4 to the left-hand operand *a*.

### 1.3 Basic expressions

Before we get into writing any expressions, we should address how we can get the bits and pieces we need.  A *function* is a block of code that only *runs* (does things) when it is called *written down*.  They are basically machines - you put in *variables* and it gives an *output* that may itself be a variable.  You can write a function, but this may be tedious, so programmers will often import *modules* which are *libraries* containing multiple functions.

Let's look at a simple example.  The *math* module contains a simple function *sqrt* that gives the square root of an *input* variable.  We can thus write a basic *expression* to determine the square root of an input variable.

In [51]:
import math       # This imports the math module
q = math.sqrt(19) # We use the equality assignment operator to assign the resulting value of math.sqrt(19) to the operand q
print(q)          # Just writing 'q' won't actually give a result; we must instead print it

4.358898943540674


**Example:** A ball is dropped from a tower of height *h* which is built on level ground. The ball has zero initial velocity and accelerates downwards under gravity. Write a program that asks the user to enter the height of the tower in metres and the time interval *t* in seconds, then prints on the screen the height of the ball from the ground at time *t* after it is dropped. Ignore air resistance.

As we know, the *print()* function displays an output. We can provide an input location using the *input()* function; this way, we don't have to keep editing our code itself to simply change a variable.  However, this will require the *float()* function, which is used to return a *float* from an *integer* or a *string*.  Using a number would simply remove all decimal values, giving an incorrect result, whilst a string is a sequenced variable type and simply could not be processed by the various functions we shall be using.

The distance from the ground is *h − gt* where $g = 9.81 m s^{-2}$ is the acceleration due to gravity. We can express this using the *round()* function, rounded to two decimal places within the bracket. As for how the answer is displayed, we can combine the actual calculation with the *round()* function within the *print()* function.  As you can see, the actual mathematical operation here is simple - it's the inputs and outputs that are somewhat complicated.  The full code is below.

In [62]:
import math
h=float(input("Enter height"))
t=float(input("Enter time interval"))
print(f"Height after time interval {t}s = {round(h-(9.81*t), 3)} m")

Enter height50
Enter time interval2.5
Height after time interval 2.5s = 25.475 m


This is a great example of why you should keep track of any brackets used in your expression.  Here, the () bracket indicates where you should put input variables; functions can be 'nested', meaning they can be put inside other functions - just see the *float(input())* part.  As for the {} brackets, they are used within the *print()* function to indicate that *h* and *t* are variables rather than just letters.  Finally, the comma used within the *round* function is because the *round()* function takes two input variables or *arguments* - the value we intend to round, and the number of decimal places which we intend to round to.  For instance, *round(5.76543, 2) = 5.77*.

The below challenge - the first available for you to try - is a great example of how code can be used in many fields aside from particle physics.

**Challenge 1:** A satellite is launched into a circular orbit around Earth such that it orbits once every *T* seconds.
- Show that the altitude h above the Earth’s surface of the satellite is $h = \left( \frac{GMT^2}{4\pi^2} \right)^{1/3} - R$ where $G = 6. 67 × 10^{-11} m^3 kg^{-1} s^{-2}$ is Newton’s gravitational constant, $M = 5. 97 × 10^{24} kg$ is the mass of the Earth and $R = 6371 km$ is the radius of Earth
- Write a program that asks the user to enter a desired value of *T* and then calculates and prints out the correct altitude in metres
- Use your program to calculate the altitudes of satellites that orbit Earth once a day (period of 1,440 minutes), once every 90 minutes and once every 45 minutes. What do you conclude from the last of these calculations?

Hint: Kepler's third law of planetary motion is given by $T^2 = \frac{4\pi^2}{GM}a^3$ where *a* is the distance from the centre of gravity.  Input your units in minutes but convert them to seconds by multiplying by 60.  You will only need the functions *float()*, *input()* and *print()*; the rest is just mathematical operators.  You may also find it easier to express the term $\frac{GMT^2}{4\pi^2}$ as a single variable - perhaps you could call it *x*?

## Challenge solutions

### Challenge 1, part 1

$T^2 = \frac{4\pi^2}{GM}a^3$ $\Longrightarrow$ $a^3 = \frac{GMT^2}{4\pi^2}$ $\Longrightarrow$ $a = \left( \frac{GMT^2}{4\pi^2} \right)^{1/3}$ $\Longrightarrow$ $h + R = \left( \frac{GMT^2}{4\pi^2} \right)^{1/3}$ $\Longrightarrow$ $h = \left( \frac{GMT^2}{4\pi^2} \right)^{1/3} - R$

### Challenge 1, part 2

The first step is to define our constants using strings:

In [2]:
G = 6.67 * 10**-11
M = 5.97 * 10**24
R = 6.371 * 10**6

We can define the time in minutes using the *input()* and *float()* functions and multiplying by 60:

In [9]:
T = float(input("time period: "))*60

time period: 1440


Now we write our law:

In [10]:
import math

x = (G * M * (T**2))/(4 * (math.pi**2))
h = x**(1/3) - R

Finally, we convert *h* to kilometres and print the result with units:

In [11]:
alt = round((h/1000),3)
print(alt, " km")

35855.91  km


As you can see, each code cell actually remembers the above contents, to an extent.  However, it would be best to simply compile our code and run it in a single cell:

In [14]:
import math

G = 6.67 * 10**-11
M = 5.97 * 10**24
R = 6.371 * 10**6

T = float(input("time period: "))*60
x = (G * M * (T**2))/(4 * (math.pi**2))
h = x**(1/3) - R
alt = round((h/1000),3)
print(alt, " km")

time period: 45
-2181.56  km


It's good habit to separate multiple lines of code into their respective purposes.  Here, the first section of code is to import the relevant modules (in this case, the *math* module), the second section is to define the necessary constants, and the third section is to give the process carried out by the code itself.

### Challenge 1, part 3

A time period of 1 day or 1,440 minutes gives an altitude of 35,855.91 km; this is the altitude of *geostationary orbit* around Earth, where an orbiting body is above the same point continually.  A period of 90 minutes gives an altitude of 279.322 km, which is a Low Earth Orbit or LEO.  Finally, a period of 45 minutes gives an altitude of -2,181.56 km - this means our satellite would actually be orbiting below the Earth's surface!